Catch warnings from sub processes during tests

Martchus opened the pull request:
See the message of the first commit for details and an example how it looks like in the error case. I'm not sure whether this is a clear win. The tests will be only aborted as a consequence of the child not being available. If the sub process is actually supposed to "passively do its thing" like the scheduler it might not immediately abort the test and the reason for the failure wouldn't be as easy to spot than in my example. Maybe it is better to explicitly handle `SIGCHLD` and call `BAIL_OUT` if it was an unexpected exit.

okurz reviewed (approved):
I like it.

> Maybe it is better to explicitly handle SIGCHLD and call BAIL_OUT if it was an unexpected exit.

Wouldn't that be a good in general?

Martchus commented:
I've updated the commit to handle `SIGCHLD`:
```
* This change will terminate sub processes spawned via our test helpers as
  soon as a warning occurs and causes the test to fail if that happens:
  [error] [pid:21734] Stopping openqa-webapi process because a Perl warning occurred: Use of uninitialized value in pattern match (m//) at /hdd/openqa-devel/repos/openQA/lib/OpenQA/WebAPI/Controller/Admin/Workers.pm line 28.
  not ok 1 - sub process openqa-webapi terminated with exit code 54784
  #   Failed test 'sub process openqa-webapi terminated with exit code 54784'
  #   at /hdd/openqa-devel/repos/openQA/t/ui/../lib/OpenQA/Test/Utils.pm line 243.
* The test will stop immediately at this point because without the sub
  process it usually makes no sense to continue further testing.
```

The code looks very hacky so but I added at least some explaining comments. By the way, I've just read the IPC::Run documentation and code and I'm really wondering why you like it. There seems not even be an easy/documented way of obtaining the PID.

okurz reviewed (commented):
woah, this turned out to be complicated but still impressive :)

Let's see what others say.

Regarding IPC::Run I consider it better than both more low-level handling as well as Mojo::IOLoop::ReadWriteProcess and I like the idea to handle foreground, external processes same as background ones. What would you prefer instead?

Martchus commented:
I prefer low-level handling over *bad* abstractions. And IPC::Run seems bad considering the code required to get the PID from a process spawned by it. I've also read the code and it seems to do a lot and it is hard to tell whether it is good or not but I had a bad feeling when reading the code. The API and its documentation also seems way worse compared to other libraries in that domain, e.g. Boost.Process or QProcess from Qt.

okurz commented:
> […] The API and its documentation also seems way worse compared to other libraries in that domain, e.g. Boost.Process or QProcess from Qt.

the question is do you find anything better from the *perl* world :)

So apparently there are no other warnings that we would have missed which is actually what I expected. Hence also codecov/patch fails as that warning handling code is never called.

Care to add tests for the part in t/lib ?

Martchus commented:
@okurz I've been adding a test… for the tests :-)

okurz reviewed (approved):
Nicely done. https://app.circleci.com/pipelines/github/os-autoinst/openQA/4352/workflows/824e4cc0-37d5-4274-a955-66d32298a5dc/jobs/41802 shows the correct execution of the new test module t/01-test-utilities.t

Martchus commented:
I couldn't get rid of the `A context appears to have been destroyed without first calling release().…` error message. The error message is likely caused by the `exit` call within `_fail_and_exit` which doesn't clean up "extra hubs on the stack" (apparently some data structures internally used by Perl's test framework). The problem is that without the `exit` call the test won't stop at this point. I tried to call `BAIL_OUT` or `die` from within the signal handler but it had no effect. Selenium tests would stop anyways due to the `BAIL_OUT` in the Selenium error handler but that doesn't help in other cases. So for now I would merge it despite that error message.

okurz reviewed (approved):
Agreed
